Задание 1 (на сопоставление)

In [23]:
import scipy.stats as stats
import math
import numpy as np
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols

In [24]:
'''Ваш ответ на задание №1 (в формате 1-А, 2-В и т.д.) '''

'''
1 - E
2 - B
3 - F, C
4 - G, D
5 - A
'''

'\n1 - E\n2 - B\n3 - F, C\n4 - G, D\n5 - A\n'

Задание 2 (про порошок)

In [25]:
'''Ваш ответ на задание №2 
- верно указана применяемая статистика и обоснование ее применения
- получено расчетное значение статистики
- получено критическое значение статистики
- получено значение p_value
- сделан статистический вывод на основе сравнения критического значения с расчетным
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- сделан содержательный вывод по поставленной проблеме'''
mean = 3.005  
true_mean = 3.00 
std = 0.015
n = 50
alpha = 0.05

z_statistic = (mean - true_mean) / (std / math.sqrt(n))

z_critical = stats.norm.ppf(1 - alpha / 2)
reject_null = abs(z_statistic) > z_critical
p_value = 2 * (1 - stats.norm.cdf(abs(z_statistic)))

print('Расчетная z-статистика: ', z_statistic)
print('Критичная z-статистика: ', z_critical)
if abs(z_statistic) > z_critical:
    print('Сравнение z-статистики с ее критичным уровнем говорит о том, что нулевая гипотеза отвергается')
elif abs(z_statistic)<=z_critical:
    print('Сравнение z-статистики с ее критичным уровнем говорит о том, что нулевая гипотеза принимается')
print('p-value: ', p_value)
print('alpha: ', alpha)
if p_value<alpha:
    print('Сравнение p-value с уровнем значимости говорит о том, что нулевая гипотеза отвергается')
elif p_value>=alpha:
    print('Сравнение p-value с уровнем значимости говорит о том, что нулевая гипотеза принимается')

Расчетная z-статистика:  2.3570226039551083
Критичная z-статистика:  1.959963984540054
Сравнение z-статистики с ее критичным уровнем говорит о том, что нулевая гипотеза отвергается
p-value:  0.01842212545410149
alpha:  0.05
Сравнение p-value с уровнем значимости говорит о том, что нулевая гипотеза отвергается


**Выводы:**

Используем z-статистику для данного распределения, т. к. у нас известно отклонение, которое задал производитель. Далее, исходя из расчитаной z-статистики можно говорить о том, что нулевая гипотеза *отвергается* и дозатор совершает ошибки. Об этом также говорит p_value, который меньше уровня значимости (который составляет *0.05*). 

Задание 3 (про двигатели)

In [26]:
'''Ваш ответ на задание №3 
- верно указана применяемая статистика и обоснование ее применения
- получено расчетное значение статистики
- получено критическое значение статистики
- получено значение p_value
- сделан статистический вывод на основе сравнения критического значения с расчетным
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- сделан содержательный вывод по поставленной проблеме'''

# Steel data
mean_steel = 90 
std_steel= 2.3 
n_steel = 15 

#Aluminum data
mean_aluminum = 88 
std_aluminum = 2.2 
n_aluminum = 10

alpha = 0.05  
df = n_steel+n_aluminum-2

# t-criteria
t_statistic = (mean_steel-mean_aluminum) / \
    (math.sqrt((std_steel**2/n_steel) + (std_aluminum**2/n_aluminum)))

t_critical = stats.t.ppf(1 - alpha / 2, df=df)
p_value = 2 * stats.t.sf(abs(t_statistic), df=df)


print('t-критерий рассчитанный: ', t_statistic)
print('t-критерий критический: ', t_critical)
if t_statistic>t_critical:
    print('Так как t-критерий больше критического значения, нулевая гипотеза отвергается')
elif t_statistic<=t_critical:
    print('Так как t-критерий меньше критического значения, нулевая гипотеза принимается')
print('p-value: ', p_value)
print('alpha: ', alpha)
if p_value<alpha:
    print('Так как p-value меньше alpha, то нулевая гипотеза отвергается')
elif p_value>=alpha:
    print('Так как p-value больше alpha, то нулевая гипотеза принимается')

t-критерий рассчитанный:  2.186521551237011
t-критерий критический:  2.0686576104190406
Так как t-критерий больше критического значения, нулевая гипотеза отвергается
p-value:  0.0392101535889427
alpha:  0.05
Так как p-value меньше alpha, то нулевая гипотеза отвергается


**Выводы:**

1) Использовалась t-статистика, т. к. необходимо было найти различие между двумя выборками;
2) Из анализа видно, что p-value и t-критерий демонстрируют, что нулевая гипотеза отвергается, что говорит о том, что существует статистически значимая разница между средними прочностями стали и алюминия.

Задание 4 (про токи)

In [27]:
'''Ваш ответ на задание №4 
- верно указана применяемая статистика и обоснование ее применения
- получено расчетное значение статистики
- получено критическое значение статистики
- получено значение p_value
- сделан статистический вывод на основе сравнения критического значения с расчетным
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- сделан содержательный вывод по поставленной проблеме'''

old_arr = np.array([80.1, 82.3, 84.1, 82.6, 85.3, 81.3, 83.2, 81.7, 82.2, 81.4])
new_arr = np.array([80.7, 81.3, 84.6, 81.7, 86.3, 84.3, 83.7, 84.7, 82.8, 84.4, 85.2, 84.9])

n1 = len(old_arr)
n2 = len(new_arr)

mean_old = np.mean(old_arr)
mean_new = np.mean(new_arr)

std_old = np.std(old_arr, ddof=1)
std_new = np.std(new_arr, ddof=1)

alpha=0.05
df = n1 + n2 - 2

var = ((n1 - 1) * std_old**2 + (n2 - 1) * std_new**2) / (n1 + n2 - 2)

SE = np.sqrt(var * (1 / n1 + 1 / n2))

t_statistic = (mean_old - mean_new) / SE
t_critical = stats.t.ppf(1 - alpha/2, df=df)

p_value = 2 * stats.t.sf(abs(t_statistic), df=df)

print('t-критерий рассчитанный: ', t_statistic)
print('t-критерий критический: ', t_critical)
if t_statistic>t_critical:
    print('Так как t-критерий больше критического значения, нулевая гипотеза отвергается')
elif t_statistic<=t_critical:
    print('Так как t-критерий меньше критического значения, нулевая гипотеза принимается')
print('p-value: ', p_value)
print('alpha: ', alpha)
if p_value<alpha:
    print('Так как p-value меньше alpha, то нулевая гипотеза отвергается')
elif p_value>=alpha:
    print('Так как p-value больше alpha, то нулевая гипотеза принимается')

t-критерий рассчитанный:  -1.8657246011714124
t-критерий критический:  2.0859634472658364
Так как t-критерий меньше критического значения, нулевая гипотеза принимается
p-value:  0.07681685091779608
alpha:  0.05
Так как p-value больше alpha, то нулевая гипотеза принимается


# Выводы
1) Использовался t-критерий, т. к. надо было найти разницу в двух выборках;
2) В результате анализа было выявлено, что нулевая гипотеза принимается, а, следовательно, разницы между средними математическими ожиданиями данных величин нет. Поэтому можно заменить старую конструкцию схемы на новую, т. к. она пропускает тот же средний ток.

Задание 5 (про глюкозу)

In [28]:
'''Ваш ответ на задание №5 
- верно указана применяемая статистика и обоснование ее применения
- получено расчетное значение статистики
- получено критическое значение статистики
- получено значение p_value
- сделан статистический вывод на основе сравнения критического значения с расчетным
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- сделан содержательный вывод по поставленной проблеме'''
before = np.array([9.6, 8.1, 8.8, 7.9, 9.2, 8.0, 8.4, 10.1, 7.8, 8.1])  # Уровень глюкозы до препарата
after = np.array([7.7, 6.2, 7.4, 7.5, 8.3, 6.2, 8.1, 8.9, 7.4, 7.0])    # Уровень глюкозы после препарата

n=len(before)
delta = before - after
alpha=0.05

mean_delta = sum(delta)/n
var = np.sqrt(sum((mean_delta-delta)**2)/(n-1))

t_statistic = mean_delta/(var/np.sqrt(n))
t_critical = stats.t.ppf(1 - alpha, df=n-1)

p_value = stats.t.sf(abs(t_statistic), df=n-1)

# Результаты
print('t-критерий рассчитанный: ', t_statistic)
print('t-критерий критический: ', t_critical)
if t_statistic>t_critical:
    print('Так как t-критерий больше критического значения, нулевая гипотеза отвергается')
elif t_statistic<=t_critical:
    print('Так как t-критерий меньше критического значения, нулевая гипотеза принимается')
print('p-value: ', p_value)
print('alpha: ', alpha)
if p_value<alpha:
    print('Так как p-value меньше alpha, то нулевая гипотеза отвергается')
elif p_value>=alpha:
    print('Так как p-value больше alpha, то нулевая гипотеза принимается')


t-критерий рассчитанный:  5.713032381721968
t-критерий критический:  1.8331129326536333
Так как t-критерий больше критического значения, нулевая гипотеза отвергается
p-value:  0.0001447470198376749
alpha:  0.05
Так как p-value меньше alpha, то нулевая гипотеза отвергается


# Выводы
1) Использовали парный t-критерий ввиду того, что исследовались 2 зависимые выборки;
2) В ходе исследования видно, что нулевая гипотеза отвергается, следовательно существуют статистически значимые различия между двумя выборками, что говорит о том, что эффект от приема препаратов присутствует.

Задание 6 (про тросы)

In [29]:
'''Ваш ответ на задание №6 
- верно указана применяемая статистика и обоснование ее применения
- получено расчетное значение статистики
- получено критическое значение статистики
- получено значение p_value
- сделан статистический вывод на основе сравнения критического значения с расчетным
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- сделан содержательный вывод по поставленной проблеме'''

new_cable = np.array([92.7, 91.6, 94.7, 93.7, 96.5, 94.5, 93.7, 96.8, 98.9, 99.9])
old_cable = np.array([90.1, 92.2, 94.3, 92.1, 93.4, 91.1, 93.2, 91.5])

alpha=0.01
df = len(new_cable)+len(old_cable)-2

t_statistic, p_value = stats.ttest_ind(new_cable, old_cable, equal_var=False, alternative='greater')
t_critical = stats.t.ppf(1 - alpha/2, df=df)

print('t-критерий рассчитанный: ', t_statistic)
print('t-критерий критический: ', t_critical)
if t_statistic>t_critical:
    print('Так как t-критерий больше критического значения, нулевая гипотеза отвергается')
elif t_statistic<=t_critical:
    print('Так как t-критерий меньше критического значения, нулевая гипотеза принимается')
print('p-value: ', p_value)
print('alpha: ', alpha)
if p_value<alpha:
    print('Так как p-value меньше alpha, то нулевая гипотеза отвергается')
elif p_value>=alpha:
    print('Так как p-value больше alpha, то нулевая гипотеза принимается')



t-критерий рассчитанный:  3.14746442193487
t-критерий критический:  2.920781622496036
Так как t-критерий больше критического значения, нулевая гипотеза отвергается
p-value:  0.003591790145908608
alpha:  0.01
Так как p-value меньше alpha, то нулевая гипотеза отвергается


# Выводы
1) Использовался независимый t-критерий, т. к. в задаче говорилось о различие дисперсии для разных видов тросов;
2) В результате видно, что нулевая гипотеза отвергается, что говорит о том, что мат. ожидания не равны, что говорит о том, что новые тросы более прочные

Задание 7 (про абитуриентов)

In [30]:
'''Ваш ответ на задание №7 
- верно указана применяемая статистика и обоснование ее применения
- получено расчетное значение статистики
- получено критическое значение статистики
- получено значение p_value
- сделан статистический вывод на основе сравнения критического значения с расчетным
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- сделан содержательный вывод по поставленной проблеме'''

before_courses = np.array([7, 6, 5, 4, 6, 2, 10, 3, 8, 5])
after_courses = np.array([9, 6, 4, 5, 7, 4, 10, 6, 9, 6])
alpha=0.01

t_statistic, p_value = stats.ttest_rel(after_courses, before_courses, alternative='greater')
t_critical = stats.t.ppf(1 - alpha, df=n-1)

print('t-критерий рассчитанный: ', t_statistic)
print('t-критерий критический: ', t_critical)
if t_statistic>t_critical:
    print('Так как t-критерий больше критического значения, нулевая гипотеза отвергается')
elif t_statistic<=t_critical:
    print('Так как t-критерий меньше критического значения, нулевая гипотеза принимается')
print('p-value: ', p_value)
print('alpha: ', alpha)
if p_value<alpha:
    print('Так как p-value меньше alpha, то нулевая гипотеза отвергается')
elif p_value>=alpha:
    print('Так как p-value больше alpha, то нулевая гипотеза принимается')

t-критерий рассчитанный:  2.7386127875258306
t-критерий критический:  2.821437925025808
Так как t-критерий меньше критического значения, нулевая гипотеза принимается
p-value:  0.011449747275884138
alpha:  0.01
Так как p-value больше alpha, то нулевая гипотеза принимается


# Выводы
1) Использовался парный t-критерий, т. к. выборки зависимы друг от друга;
2) В результате анализа видно, что t-критерий меньше критического значения, следовательно, нулевая гипотеза принимается, следовательно, средние результаты тестов одинаковы, что значеит, что подготовительные курсы не повлияли на абитуриентов

Задание 8 (про фрукты и IQ)

In [31]:
'''Ваш ответ на задание №8 
- верно указана применяемая статистика и обоснование ее применения
- получено расчетное значение статистики
- получено критическое значение статистики
- получено значение p_value
- сделан статистический вывод на основе сравнения критического значения с расчетным
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- сделан содержательный вывод по поставленной проблеме'''

orange_group = [125, 138, 112, 130, 124, 125, 110, 112, 117, 126]
pear_group = [124, 102, 142, 130, 131, 124, 127, 115, 108, 109]
alpha = 0.05

var_group_orange = np.var(orange_group, ddof=1)
var_group_pear = np.var(pear_group, ddof=1)

if var_group_orange>var_group_pear:
    F_statistic = var_group_orange / var_group_pear 
else:
    F_statistic = var_group_pear / var_group_orange

df_group_1 = len(orange_group) - 1
df_group_2 = len(pear_group) - 1

alpha = 0.05
F_critical = stats.f.ppf(1 - alpha / 2, df_group_1, df_group_2)
p_value = 2 * stats.f.sf(abs(F_statistic), df_group_1, df_group_2)

print('F-критерий рассчитанный: ', F_statistic)
print('F-критерий критический: ', F_critical)
if F_statistic>F_critical:
    print('Так как F-критерий больше критического значения, нулевая гипотеза отвергается')
elif F_statistic<=F_critical:
    print('Так как F-критерий меньше критического значения, нулевая гипотеза принимается')
print('p-value: ', p_value)
print('alpha: ', alpha)
if p_value<alpha:
    print('Так как p-value меньше alpha, то нулевая гипотеза отвергается')
elif p_value>=alpha:
    print('Так как p-value больше alpha, то нулевая гипотеза принимается')


F-критерий рассчитанный:  1.9061769156692807
F-критерий критический:  4.025994158282979
Так как F-критерий меньше критического значения, нулевая гипотеза принимается
p-value:  0.35058724116448386
alpha:  0.05
Так как p-value больше alpha, то нулевая гипотеза принимается


# Выводы
1) Был использован F-критерий Фишера, т. к. необходимо было произвести сравнение дисперсий двух выборок;
2) В результате анализа было нулевая гипотеза была принята, что говорит о том, что дисперсии данных двух групп одинаковые.

Задание 9 (про самолеты)

In [32]:
'''Ваш ответ на задание №9 
- верно указана применяемая статистика и обоснование ее применения
- получено расчетное значение статистики
- получено критическое значение статистики
- получено значение p_value
- сделан статистический вывод на основе сравнения критического значения с расчетным
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- сделан содержательный вывод по поставленной проблеме'''

flights = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9])
frequencies = np.array([36, 23, 11, 13, 8, 3, 2, 1, 3])

mean = sum(flights*frequencies)/sum(frequencies)

p_value = 1 / (1 + mean)  
geom_probs = stats.geom.pmf(flights, p_value)
geom_probs[-1] += 1 - stats.geom.cdf(9, p_value)
expected_geom = sum(frequencies) * geom_probs
chi2_critical = stats.chi2.ppf(1 - alpha, df=len(frequencies)-2)
chi2_stat = np.sum((frequencies - expected_geom) ** 2 / expected_geom)

print('Хи-квадрат рассчитанный: ', chi2_stat)
print('Хи-квадрат критический: ', chi2_critical)
if chi2_stat>chi2_critical:
    print('Так как Хи-квадрат больше критического значения, нулевая гипотеза отвергается')
elif chi2_stat<=chi2_critical:
    print('Так как Хи-квадрат меньше критического значения, нулевая гипотеза принимается')
print('p-value: ', p_value)
print('alpha: ', alpha)
if p_value<alpha:
    print('Так как p-value меньше alpha, то нулевая гипотеза отвергается')
elif p_value>=alpha:
    print('Так как p-value больше alpha, то нулевая гипотеза принимается')


Хи-квадрат рассчитанный:  12.266673591341583
Хи-квадрат критический:  14.067140449340167
Так как Хи-квадрат меньше критического значения, нулевая гипотеза принимается
p-value:  0.267379679144385
alpha:  0.05
Так как p-value больше alpha, то нулевая гипотеза принимается


# Вывод
1) Было найдено *геометрической* распределение, при которой нулевая гипотеза принимается (т. е. распределение соответствует заданному). Также исследовалось распределение Пуассона, которое дало негативный результат;
2) Использовалась статистика chi2 для данного вывода.

Задание 10 (про цвет глаз и волос)

In [33]:
'''Ваш ответ на задание №10
- верно указана применяемая статистика и обоснование ее применения
- получено расчетное значение статистики
- получено критическое значение статистики
- получено значение p_value
- сделан статистический вывод на основе сравнения критического значения с расчетным
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- сделан содержательный вывод по поставленной проблеме'''

df = pd.DataFrame({
    "Светлые": [35, 28, 7],
    "Темные": [52, 42, 10],
    "Рыжие": [13, 10, 3]
}, index=["Голубые", "Карие", "Зеленые"])
alpha = 0.05

chi2_stat, p_value, dof, expected = stats.chi2_contingency(df)
chi2_critical = stats.chi2.ppf(1 - alpha, df=dof)

print('Хи-квадрат рассчитанный: ', chi2_stat)
print('Хи-квадрат критический: ', chi2_critical)
if chi2_stat>chi2_critical:
    print('Так как Хи-квадрат больше критического значения, нулевая гипотеза отвергается')
elif chi2_stat<=chi2_critical:
    print('Так как Хи-квадрат меньше критического значения, нулевая гипотеза принимается')
print('p-value: ', p_value)
print('alpha: ', alpha)
if p_value<alpha:
    print('Так как p-value меньше alpha, то нулевая гипотеза отвергается')
elif p_value>=alpha:
    print('Так как p-value больше alpha, то нулевая гипотеза принимается')


Хи-квадрат рассчитанный:  0.09615384615384615
Хи-квадрат критический:  9.487729036781154
Так как Хи-квадрат меньше критического значения, нулевая гипотеза принимается
p-value:  0.9988806869095861
alpha:  0.05
Так как p-value больше alpha, то нулевая гипотеза принимается


# Выводы
1) Для нахождения взаимосвязи между цветом волос и глаз использовался критерий независимости chi2;
2) Исходя из того, что нулевая гипотеза говорит о том, что переменные не связанны друг с другом можно говорить о том, что взаимосвязи между цветом глаз и цветом волос выявлено не было.

Задание 11 (про тигров)

In [34]:
'''Ваш ответ на задание №11
- верно указана применяемая статистика и обоснование ее применения
- получено расчетное значение статистики
- получено значение p_value
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- в случае, если нулевая гипотеза отвергнута, сделан тест Тьюки для подробного сравнения средних
- сделано дополнение к статистическому выводу по результатам теста Тьюки
- сделан содержательный вывод по поставленной проблеме'''
tukey_test=0
alpha=0.05
data = [[151, 135, 137, 118, 132, 135, 131, 121, 137, 140, 152, 133, 151, 132],
               [108, 123, 145, 103, 110, 124, 116, 108, 104, 130, 106, 105, 114, 108],
               [140, 152, 149, 132, 124, 146, 156, 137, 124, 145, 147, 139, 157, 139],
               [127, 135, 110, 116, 119, 123, 102, 121, 118, 120, 114, 134, 121, 109]]
    
anova_statistic, p_value = stats.f_oneway(*data)
anova_critical = stats.f.ppf(1-alpha, len(data)-1, (len(data[0])*4)-len(data))
print('ANOVA рассчитанный: ', anova_statistic)
print('ANOVA критический: ', anova_critical)
if anova_statistic>anova_critical:
    print('Так как ANOVA больше критического значения, нулевая гипотеза отвергается')
elif anova_statistic<=anova_critical:
    print('Так как ANOVA меньше критического значения, нулевая гипотеза принимается')
print('p-value: ', p_value)
print('alpha: ', alpha)
if p_value<alpha:
    print('Так как p-value меньше alpha, то нулевая гипотеза отвергается')
    tukey_test=1
elif p_value>=alpha:
    print('Так как p-value больше alpha, то нулевая гипотеза принимается')

if tukey_test==1:
    print('ANOVA показал, что выборки различны. По условию задания необходимо ' +
          'найти пару (или пары) с различиями в среднем значении с помощью теста Тьюки')
    
res = stats.tukey_hsd(*data)
print(res)

ANOVA рассчитанный:  21.927813243053045
ANOVA критический:  2.7826004234346047
Так как ANOVA больше критического значения, нулевая гипотеза отвергается
p-value:  2.592780238646728e-09
alpha:  0.05
Так как p-value меньше alpha, то нулевая гипотеза отвергается
ANOVA показал, что выборки различны. По условию задания необходимо найти пару (или пары) с различиями в среднем значении с помощью теста Тьюки
Tukey's HSD Pairwise Group Comparisons (95.0% Confidence Interval)
Comparison  Statistic  p-value  Lower CI  Upper CI
 (0 - 1)     21.500     0.000    10.981    32.019
 (0 - 2)     -5.857     0.458   -16.376     4.662
 (0 - 3)     16.857     0.000     6.338    27.376
 (1 - 0)    -21.500     0.000   -32.019   -10.981
 (1 - 2)    -27.357     0.000   -37.876   -16.838
 (1 - 3)     -4.643     0.647   -15.162     5.876
 (2 - 0)      5.857     0.458    -4.662    16.376
 (2 - 1)     27.357     0.000    16.838    37.876
 (2 - 3)     22.714     0.000    12.195    33.233
 (3 - 0)    -16.857     0.000 

# Вывод
1) Для нахождения разницы в средних значениях был использован дисперсионный анализ (т. к. количество групп больше 2), в результате которого было выяснено, что присутствуют различия между средними величинами;
2) После этого был применен тест Тьюки и были найдены пары групп с равной средней величиной: 1 и 3 (с p_value равным 0.458), а также 2 и 4 (c p_value равным 0.647);
3) В остальных группах средние величины не равны.

Задание 12 (про белок)

In [35]:
'''Ваш ответ на задание №12
- верно указана применяемая статистика и обоснование ее применения
- расчет выполнен через последовательные суммы квадратов
- получены расчетные значения статистики
- получены значения p_value
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- сделан содержательный вывод по поставленной проблеме'''

'Ваш ответ на задание №12\n- верно указана применяемая статистика и обоснование ее применения\n- расчет выполнен через последовательные суммы квадратов\n- получены расчетные значения статистики\n- получены значения p_value\n- сделан статистический вывод на основе сравнения p-value и уровня значимости\n- сделан содержательный вывод по поставленной проблеме'

In [36]:
data = [[440, 438, 429, 502, 602], [308, 328, 326, 326, 325],
        [892, 868, 855, 866, 932], [737, 798, 876, 810, 861],
        [1575, 849, 759, 1602, 1327], [1001, 901, 958, 1032, 883]]
alpha = 0.05

data_groups = [np.array(values) for values in data]
group_len = [len(values) for values in data]

total_mean = np.mean([value for group in data_groups for value in group])
group_means = [np.mean(values) for values in data_groups]

SST = sum((value - total_mean) ** 2 for group in data_groups for value in group)
SSR = sum(n * np.sum((values - np.mean(values)) ** 2)
                for n, values in zip(group_len, data_groups))
SSE= sum(n * (mean - total_mean) ** 2 for n, mean in zip(group_len, group_means))

df_between = len(data_groups) - 1
df_within = sum(group_len) - len(data_groups)

F_statistic = (SSE / df_between) / (SSR / df_within)

F_critical = stats.f.ppf(1 - alpha, df_between, df_within)
p_value = stats.f.sf(F_statistic, df_between, df_within)

print('F-критерий рассчитанный: ', F_statistic)
print('F-критерий критический: ', F_critical)
if F_statistic>F_critical:
    print('Так как F-критерий больше критического значения, нулевая гипотеза отвергается')
elif F_statistic<=F_critical:
    print('Так как F-критерий меньше критического значения, нулевая гипотеза принимается')
print('p-value: ', p_value)
print('alpha: ', alpha)
if p_value<alpha:
    print('Так как p-value меньше alpha, то нулевая гипотеза отвергается')
elif p_value>=alpha:
    print('Так как p-value больше alpha, то нулевая гипотеза принимается')


F-критерий рассчитанный:  3.7441644909072074
F-критерий критический:  2.620654147862887
Так как F-критерий больше критического значения, нулевая гипотеза отвергается
p-value:  0.012001943352027247
alpha:  0.05
Так как p-value меньше alpha, то нулевая гипотеза отвергается


# Выводы
1) Расчитаны общая, объясненная и остаточная суммы квадратов. Исходя из них расчитан F-критерий, который показывает, что нулевая гипотеза отвергается, что свидетельствует о статистически значимых различиях в наблюдаемых группах.

In [37]:
# NEW VERSION
df = pd.DataFrame({
    "Возраст": ["1 год"] * 5 + ["2 года"] * 5 + ["3 и более лет"] * 5 + ["1 год"] * 5 + ["2 года"] * 5 + ["3 и более лет"] * 5,
    "Выводок": ["Без выводка"] * 15 + ["С выводком"] * 15,
    "Масса": [
        440, 438, 429, 502, 602,  # Без выводка, 1 год
        892, 868, 855, 866, 932,  # Без выводка, 2 года
        1575, 849, 759, 1602, 1327,  # Без выводка, 3 и более лет
        308, 328, 326, 326, 325,  # С выводком, 1 год
        737, 798, 876, 810, 861,  # С выводком, 2 года
        1001, 901, 958, 1032, 883  # С выводком, 3 и более лет
    ]
})

model = ols('Масса ~ C(Возраст) + C(Выводок) + C(Возраст):C(Выводок)', data=df).fit()
anova_table = sm.stats.anova_lm(model, typ=2)

print(anova_table)

                          sum_sq    df          F        PR(>F)
C(Возраст)             2427075.8   2.0  42.377707  1.333903e-08
C(Выводок)              202705.2   1.0   7.078626  1.368668e-02
C(Возраст):C(Выводок)    50688.2   2.0   0.885036  4.257428e-01
Residual                687269.6  24.0        NaN           NaN


# Выводы 2
1) Результаты многофакторного дисперсионного анализа показывают, что влияние фактора возраста имеет знеачение на массу тела (*p-value* равен 1.33\*10^-8, что меньше 0.05), как и фактор наличия выводка (*p-value* равен 0.0136, что меньше 0.05), однако совместное влияние факторов незначительно (*p-value* равен 0.425, что больше 0.05)